### ISSUE TITLE
Establish format for metadata associated with each scan

### ISSUE THREAD

**Issue by joshmutus** [Boulder-Cryogenic-Quantum-Testbed/measurement#12]
A scan, which will contain amplitude and phase data has to have some metadata indicating the configuration state around it. I wanted to get people's thoughts on what has to be included. Here's a starting point from our own data:

avg_power
bandwidth
fit_theta
freq
min_avgs
points
power
size
span
span_factor
stitch_power
time_out

@cary-earnest @crmcrae @MartinWeides suggestions for other parameters to include? Things like temperature would be great to have, but perhaps as optional fields?

**Comment by MartinWeides** [comment]
This seems useful. Since VNA sweeps are (always?) upward in frequency the direction of scan may not need to be included. Maybe incluede temperature?

**Comment by cary-earnest** [comment]
@joshmutus can you explain more what each of those things are? e.g what is stitch_power? wouldn't the actual number of averages be more useful than the min number of averages? The number of points taken are self explanatory from the data length, is it useful to record that separately?  What is the difference between size/span/spanfactor?

**Comment by joshmutus** [comment]
@cary-earnest all good questions. I just pasted something from a dataset. I'll figure out what's going on here...

**Comment by crmcrae** [comment]
From Jeremy's email yesterday, it's clear that we need to establish a way to report the data format and units, as well as any preprocessing that may have been performed. We should add this as soon as possible to ensure that the community survey of resonator fits is accurate.

**Comment by joshmutus** [comment]
Yes. I sent Cary and Jeremy an averaged dataset to see if that is more suitable. This is a good exercise to see all the assumptions we have about the data

**Comment by jebej** [comment]
Hi all, it would make sense to have all of the input data in the same format, to avoid having everyone replicate the data input routines for all the different datasets.

It seems simplest to have the data be in the format that you end up feeding into the fitting algorithm, post averaging & any normalization/calibration. For us, that would be a single "trace", with the line attenuation calibrated away. Whether the data is complex or mag/phase is not an issue since it is trivial to convert.

The issue of normalization/calibration is important though, and maybe can be explored separately from the actual fitting. I also realize that it is possible that some groups might not pre-calibrate, and instead include the line attenuation directly into the fit. This would throw a wrench into the above.

**Comment by crmcrae** [comment]
Just chiming in to say that Leo Ranzani at BBN has also contacted me with issues fitting the IQC data due to the frequency shift. I'm copying Leo's comment below:
"We couldn't fit the Waterloo resonator, because it looks like the frequency axis has been shifted to center the resonator at 0Hz. If Matteo can provide the center frequency we could try again though (otherwise the answer is easy: Q factor is zero :)"

**Comment by crmcrae** [comment]
Hi all,

Due to community feedback about issues with fitting a variety of data formats, I went through the datasets we have so far and compiled guidelines for a standard data format. Going forward, we hope that everyone can share their data in the following format:

Three column .csv file starting in row 1, with no empty rows between data.
Data should be the raw data with no normalization or frequency shift.

**Column 1:**
Monotonically increasing frequency in GHz

**Column 2:**
Magnitude of S21 in dB (log mag)

**Column 3:**
Phase of S21 in radians

**Other info to be included in a .txt accompanying the data (encouraged but not required):**
Input power (dBm)
Number of averages
IFBW
Fridge temperature (mK)
Type of device (materials, design geometry, etc.)

Please let me know if you have any questions or feedback about the guidelines. We're still in the early stages of this effort so any comments would be very helpful.

The NIST and Google data should be in the correct format already, and the GitHub will be updated when other datasets are reformatted.

Thanks!

**Comment by crmcrae** [comment]
Oh and I'll close this thread and make a new thread to discuss changes and steps going forward with the standard data format.

**Comment by crmcrae** [comment]
The Keysight Touchstone data format doesn't include any metadata associated with the scans. How should we go about including this? What do our metadata files look like now @mullinska ?

**Comment by FaustinCarter** [comment]
Two reasonable options:
1.  Use a format like netcdf4 that allows metadata and then provide methods
to read and export touchstone files. The xarray package would help you do
this seamlessly. This would also let you store sweeps of temperature,
power, magnetic field, etc in a native highly compact binary format.

2. Put a JSON string with metadata as the first comment line in the
touchstone file. Then just read the first line in to a dict on load prior
to parsing the rest of the file with scikit-rf (no need to reinvent the
wheel).

Some combination of these is probably the best. Allow for import of
touchstone, but natively store in netcdf4. If someone wants to export to
touchstone, include metadata as a comment.

**Comment by mullinska** [comment]
@crmcrae The current metadata files (and by that I assume you mean the file that prints the conditions for the sweep ran), looks like this:
STARTPOWER: X dB
ENDPOWER: X dB
NUMSWEEPS: X
CENTERF: X GHz
SPAN: X MHz
TEMP: X mK
STARTING AVERAGES: X
EDELAY: X ns
IFBAND: X kHz
POINTS: X
